filewatcher/filewatcher.go

```diff
@@ -104,6 +105,35 @@ func (r *Result) watcherLoop(
 	}
 }
 
+// Config defines the config to be used in New function.
+type Config struct {
+	// The path to the file to be watched, required.
+	Path string
+
+	// The parser to parse the data load, required.
+	Parser Parser
+
+	// Optional. When non-nil, it will be used to log errors,
+	// either returned by parser or by the underlying file system watcher.
+	// Please note that this does not include errors returned by the first parser
+	// call, which will be returned directly.
+	Logger log.Wrapper
+
+	// Optional. When <=0 DefaultMaxFileSize will be used instead.
+	// This limits the size of the file that will be read into memory and send to
+	// the parser, to limit memory usage.
+	// If the file content is larger than MaxFileSize,
+	// we don't treat that as an error,
+	// but the parser will only receive the first MaxFileSize bytes of content.
```

Comment:
🔕 we don't _treat_ this as an error, but generally won't the parser only receiving a portion of the file _result_ in an error?  Would it be better to actually treat this as an error rather than letting it return unpredictable errors later in the execution?

Reply:
A few considerations here:

1. Some parser stops reading when they got everything they need (for example, [json](https://play.golang.org/p/QP9yudEszSd), but those parsers won't have a problem even we don't limit the data sent to them)
2. Yes for most parsers only receiving partial data will cause an error, but when a parser returns an error we won't update the parsed result, so that won't cause us problems.
3. Detect the fact that file is too large without actually reading it is tricky. You can use Lstats to get the file size, but that doesn't work for all cases (e.g. symlinks, special files like `/dev/urandom`, etc.). This is a more robust solution.

Reply:
👍 
It might be good to include those points in the comment for future reference/documentation?

Reply:
Won't this hide the underlying cause of parse failures?

Reply:
@konradreiche There's no reliable way to detect the underlying cause, though. Parser implementations can also count the bytes read to see if it's at the limit.

Reply:
I think making some effort to at least inform the user that there are still more bytes available would greatly help in case there's some misconfiguration going on.

Reply:
There's no reliable way to know that without actually reading the file either.

Reply:
I think the fact that there are edge cases in detecting the file size can be discouraging but is it enough to justify not making an attempt to detect it while it will most likely work in many cases to prevent masked bugs sneaking in?

Reply:
As long as their parser return an error, they can figure it out from the error logged/sent to raven (see also: https://github.com/reddit/baseplate.go/pull/153)

Have a unified way (that it's always a parser error) is better because if it usually is another error, but occasionally it's a parser error, then in the long run they might start to only handle it in the other error that doesn't work 100% of the time, and forgot that it could also be a parser error.